dynawo/sources/Models/CPP/Controls/Shunts/ModelCentralizedShuntsSectionControl/DYNModelCentralizedShuntsSectionControl.cpp

```diff
@@ -0,0 +1,346 @@
+    isSelf_(false),
+    URef0Pu_(0.),
+    changingShunt(-1),
+    whenUp_(VALDEF),
+    whenDown_(VALDEF),
+    tNext_(10.),
+    lastTime_(VALDEF) {}
+
+  void
+  ModelCentralizedShuntsSectionControl::defineParameters(std::vector<ParameterModeler>& parameters) {
+    parameters.push_back(ParameterModeler("nbShunts", VAR_TYPE_INT, EXTERNAL_PARAMETER));
+    parameters.push_back(ParameterModeler("section0", VAR_TYPE_INT, EXTERNAL_PARAMETER, "*", "nbShunts"));
+    parameters.push_back(ParameterModeler("SectionMin", VAR_TYPE_INT, EXTERNAL_PARAMETER, "*", "nbShunts"));
+    parameters.push_back(ParameterModeler("SectionMax", VAR_TYPE_INT, EXTERNAL_PARAMETER, "*", "nbShunts"));
+    parameters.push_back(ParameterModeler("DeadbandUPu", VAR_TYPE_DOUBLE, EXTERNAL_PARAMETER, "*", "nbShunts"));
+    parameters.push_back(ParameterModeler("IsSelf", VAR_TYPE_BOOL, EXTERNAL_PARAMETER));
+    parameters.push_back(ParameterModeler("URef0Pu", VAR_TYPE_DOUBLE, EXTERNAL_PARAMETER));
+    parameters.push_back(ParameterModeler("tNext", VAR_TYPE_DOUBLE, EXTERNAL_PARAMETER));
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::setSubModelParameters() {
+    try {
+      nbShunts_ = findParameterDynamic("nbShunts").getValue<int>();
+      isSelf_ = findParameterDynamic("IsSelf").getValue<bool>();
+      URef0Pu_ = findParameterDynamic("URef0Pu").getValue<double>();
+      tNext_ = findParameterDynamic("tNext").getValue<double>();
+      for (int s = 0; s < nbShunts_; ++s) {
+        std::stringstream section0Name;
+        section0Name << "section0_" << s;
+        sections0_.push_back(findParameterDynamic(section0Name.str()).getValue<int>());
+        std::stringstream SectionMinName;
+        SectionMinName << "SectionMin_" << s;
+        SectionsMin_.push_back(findParameterDynamic(SectionMinName.str()).getValue<int>());
+        std::stringstream SectionMaxName;
+        SectionMaxName << "SectionMax_" << s;
+        SectionsMax_.push_back(findParameterDynamic(SectionMaxName.str()).getValue<int>());
+        std::stringstream DeadbandUPuName;
+        DeadbandUPuName << "DeadbandUPu_" << s;
+        DeadbandsUPu_.push_back(findParameterDynamic(DeadbandUPuName.str()).getValue<double>());
+      }
+    } catch (const DYN::Error& e) {
+    Trace::error() << e.what() << Trace::endline;
+    throw DYNError(Error::MODELER, NetworkParameterNotFoundFor, staticId());
+    }
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::defineVariables(std::vector<boost::shared_ptr<Variable> >& variables) {
+    // only one UMonitoredPu variable shared among all shunts as all of them are regulating the same node
+    variables.push_back(VariableNativeFactory::createState("UMonitoredPu_value", CONTINUOUS));
+    for (int s = 0; s < nbShunts_; ++s) {
+      std::stringstream URefPuName;
+      URefPuName << "URefPu_" << s << "_value";
+      variables.push_back(VariableNativeFactory::createState(URefPuName.str(), DISCRETE));
+    }
+    for (int s = 0; s < nbShunts_; ++s) {
```

Comment:
could be done in the same for loop as above ?

Reply:
it is better to have same kind of variable close to each other